stan/math/opencl/kernel_generator/reduction_2d.hpp

```diff
@@ -0,0 +1,342 @@
+ * @param a expression to reduce
+ * @return sum
+ */
+template <typename T, require_all_kernel_expressions_t<T>* = nullptr>
+inline auto sum_2d(T&& a) {
+  auto&& arg_copy = as_operation_cl(std::forward<T>(a)).deep_copy();
+  return sum_2d_<as_operation_cl_t<T>>(as_operation_cl(std::forward<T>(a)));
+}
+
+/**
+ * Represents two dimensional max - reduction in kernel generator expressions.
+ * @tparam T type of expression
+ */
+template <typename T>
+class max_2d_
+    : public reduction_2d<max_2d_<T>, T,
+                          max_op<typename std::remove_reference_t<T>::Scalar>> {
+  using base
+      = reduction_2d<max_2d_<T>, T,
+                     max_op<typename std::remove_reference_t<T>::Scalar>>;
+  using base::arguments_;
+
+ public:
+  using op = max_op<typename std::remove_reference_t<T>::Scalar>;
+  explicit max_2d_(T&& a)
+      : reduction_2d<max_2d_<T>, T, op>(std::forward<T>(a), op::init()) {}
+  /**
+   * Creates a deep copy of this expression.
+   * @return copy of \c *this
+   */
+  inline auto deep_copy() const {
+    auto&& arg_copy = this->template get_arg<0>().deep_copy();
+    return max_2d_<std::remove_reference_t<decltype(arg_copy)>>(
+        std::move(arg_copy));
+  }
+};
+
+/**
+ * two dimensional max - reduction of a kernel generator expression. So as to
+ * be efficient two dimensional reductions are only done partially. That means
+ * instead of 1 element kernel output can have a few rows and a few columns that
+ * need to be reduced to obtain final result. This can be done in a separate
+ * kernel or after copying to CPU. Also two dimensional reductions can not be
+ * used as arguments to other operations - they can only be evaluated.
+ * @tparam T type of input expression
+ * @param a expression to reduce
+ * @return max
+ */
+template <typename T, require_all_kernel_expressions_t<T>* = nullptr>
+inline auto max_2d(T&& a) {
+  auto&& arg_copy = as_operation_cl(std::forward<T>(a)).deep_copy();
+  return max_2d_<as_operation_cl_t<T>>(as_operation_cl(std::forward<T>(a)));
+}
+
+/**
+ * Represents two dimensional min - reduction in kernel generator expressions.
+ * @tparam T type of expression
+ */
+template <typename T>
+class min_2d_
+    : public reduction_2d<min_2d_<T>, T,
+                          min_op<typename std::remove_reference_t<T>::Scalar>> {
+  using base
+      = reduction_2d<min_2d_<T>, T,
+                     min_op<typename std::remove_reference_t<T>::Scalar>>;
+  using base::arguments_;
+
+ public:
+  using op = min_op<typename std::remove_reference_t<T>::Scalar>;
+  explicit min_2d_(T&& a)
+      : reduction_2d<min_2d_<T>, T, op>(std::forward<T>(a), op::init()) {}
+  /**
+   * Creates a deep copy of this expression.
+   * @return copy of \c *this
+   */
+  inline auto deep_copy() const {
+    auto&& arg_copy = this->template get_arg<0>().deep_copy();
+    return min_2d_<std::remove_reference_t<decltype(arg_copy)>>(
+        std::move(arg_copy));
+  }
+};
```

Comment:
[optional] I'll leave this to you but this might be a reasonable spot to have a macro to generate these. Or do they slightly differ in some way I'm not seeing?

Reply:
We could use a macro. But it is just a few cases so I am not sure if it is worth it.